CRM/Contact/BAO/Query.php

```diff
@@ -6332,6 +6324,14 @@ protected function prepareOrderBy($sort, $sortOrder) {
       $direction = isset($orderByClauseParts[1]) ? $orderByClauseParts[1] : 'asc';
       $fieldSpec = $this->getMetadataForRealField($field);
 
+      // This is a hacky add-in for primary address joins. Feel free to iterate as it is unit tested.
+      // @todo much more cleanup on location handling in addHierarchical elements. Potentially
+      // add keys to $this->fields to represent the actual keys for locations.
+      if (empty($fieldSpec) && substr($field, 0, 2) === '1-') {
+        $fieldSpec = $this->getMetadataForField(substr($field, 2));
+        $this->addAddressTable('1-' . str_replace('civicrm_', '', $fieldSpec['table_name']), 'is_primary = 1');
+      }
```

Comment:
Have tested the complete changes and confirm that it works fine on profile search through advanced search. 

Maybe, if we extend the above condition to also support for primary location type "names" eg `Home-`, this will also fix the same problem in search builder? @eileenmcnaughton

Reply:
@jitendrapurohit yeah - do we still have that problem in search builder or is it fixed now?

I'd rather not fix anything in this area without adding test cover so if we do still have a problem I think that would be better as a follow up

Reply:
> yeah - do we still have that problem in search builder or is it fixed now?

We do have them, and it exists after this PR too. But can be fixed if the solution for `Home-` type is extended. There also seems to be other issues with address field search, eg 

Advanced search -> select search profile -> select loc type = non primary -> search -> Error.

But I concur it also errors before this PR so these changes are better in than out as it fixes the main issue with profile search sorting 👍